File: AcodeIOS/Acode/Networking/LanSubnetProbe.swift
import Foundation

enum LanSubnetProbe {
    static func discoverHealthHost(port: Int, preferredHost: String? = nil, session: URLSession = .shared) async -> String? {
        if let preferredHost, !preferredHost.isEmpty, await healthOK(host: preferredHost, port: port, session: session) {
            return preferredHost
        }
        guard let prefix = LanNetworkSelector.wifiSubnetPrefix() else { return nil }
        for host in 1...254 {
            let ip = "\(prefix).\(host)"
            if ip == preferredHost { continue }
            if await healthOK(host: ip, port: port, session: session) {
                return ip
            }
        }
        return nil
    }

    private static func healthOK(host: String, port: Int, session: URLSession) async -> Bool {
        guard let url = URL(string: "http://\(host):\(port)/health") else { return false }
        var request = URLRequest(url: url)
        request.timeoutInterval = 1
        do {
            let (_, response) = try await session.data(for: request)
            return (response as? HTTPURLResponse)?.statusCode == 200
        } catch {
            return false
        }
    }
}

File: Android/app/src/main/java/com/codevoke/android/data/LanSubnetProbe.kt
package com.codevoke.android.data

import android.content.Context
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import java.util.concurrent.TimeUnit

object LanSubnetProbe {
    suspend fun discoverHealthHost(
        context: Context,
        port: Int,
        preferredHost: String? = null,
    ): String? = withContext(Dispatchers.IO) {
        val client = LanNetworkSelector.wifiBoundClient(context, connectTimeoutSeconds = 1)
            ?: OkHttpClient.Builder()
                .connectTimeout(1, TimeUnit.SECONDS)
                .readTimeout(1, TimeUnit.SECONDS)
                .build()
        preferredHost?.takeIf { it.isNotBlank() }?.let { host ->
            if (healthOk(client, host, port)) return@withContext host
        }
        val prefix = LanNetworkSelector.wifiSubnetPrefix(context) ?: return@withContext null
        coroutineScope {
            (1..254).chunked(40).forEach { chunk ->
                val found = chunk.map { host ->
                    async {
                        val ip = "$prefix.$host"
                        if (ip == preferredHost) null else if (healthOk(client, ip, port)) ip else null
                    }
                }.awaitAll().firstOrNull { !it.isNullOrBlank() }
                if (found != null) return@coroutineScope found
            }
            null
        }
    }

    private fun healthOk(client: OkHttpClient, host: String, port: Int): Boolean =
        runCatching {
            val request = Request.Builder().url("http://$host:$port/health").build()
            client.newCall(request).execute().use { it.isSuccessful }
        }.getOrDefault(false)
}
